### PR TITLE
Enable type assertion linter

### PR DESCRIPTION
**Pull request by lindnerby**
**Description**

Changes proposed in this pull request:

- Bump golangci-lint version
- Disabled `depguard` that comes with this version bump and we have no package whitelist
- Introduce errors for failed type assertions
- Check type assertions

**Related issue(s)**
* https://github.com/kyma-project/lifecycle-manager/issues/694

### CODE REVIEW

File: api/v1beta2/error.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package v1beta2
+
+import "errors"
+
+var (
+	ErrTypeAssertKyma           = errors.New("raw can not be cast to v1beta2.Kyma")
```
Comment (Tomasz-Smelcerz-SAP) [inline api/v1beta2/error.go:6]: I don't understand what `raw` means in the generic context. Perhaps the `value can not be cast to v1beta2.Manifest` would be better?

File: pkg/common/error.go [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package common
+
+import "errors"
+
+var ErrTypeAssert = errors.New("generic TypeAssertion error")
```
Comment (Tomasz-Smelcerz-SAP) [inline pkg/common/error.go:5]: Perhaps just: "TypeAssertion failed" ?

**Comment by Tomasz-Smelcerz-SAP** [comment]
I have found a not-detected type assertion:

```
controllers/control-plane/kcp_controller_helper_test.go
110:    value, found := initKey.(map[string]any)["initKey"]
```
Why it wasn't detected by the linter?

File: internal/declarative/v2/reconciler.go [hunk 3]
```diff
@@ -463,8 +468,12 @@ func updateSyncedOCIRefAnnotation(obj Object, ref string) {
 }
 
 func pruneResource(diff []*resource.Info, resourceType string, resourceName string) []*resource.Info {
+	//nolint:varnamelen
 	for i, info := range diff {
-		obj := info.Object.(client.Object)
+		obj, ok := info.Object.(client.Object)
+		if !ok {
```
Comment (Tomasz-Smelcerz-SAP) [inline internal/declarative/v2/reconciler.go:474]: Is hiding the conversion problem here OK? Perhaps an error should be returned instead?

**Comment by lindnerby** [comment]
> I have found a not-detected type assertion:
> 
> ```
> controllers/control-plane/kcp_controller_helper_test.go
> 110:    value, found := initKey.(map[string]any)["initKey"]
> ```
> 
> Why it wasn't detected by the linter?
Guess because the map key exists check is covering it?  No idea, I separated it.

**Pull request closed without merging**